File: app/routes/pickups.py
"""
Pickup routes.

    GET  /api/pickups             — my own pickups (provider or recipient)
    GET  /api/pickups/<id>        — view one (owning provider OR owning recipient)
    POST /api/pickups/<id>/confirm  — recipient confirms a scheduled pickup
    POST /api/pickups/<id>/complete — provider marks the handover complete
    POST /api/pickups/<id>/fail     — provider marks a no-show/failed pickup
    POST /api/pickups/<id>/cancel   — recipient cancels a pickup they can't make

Scheduling a pickup is registered on `requests_bp` instead (the
resource in that URL is a request, not yet a pickup — see
`POST /api/requests/<id>/schedule-pickup` in app/routes/requests.py).
"""

from flask import Blueprint, g, request

from app.auth.decorators import role_required, token_required
from app.models import UserRole
from app.services.pickup_service import (
    PickupError,
    cancel_pickup,
    complete_pickup,
    confirm_pickup,
    fail_pickup,
    get_pickup_for_viewer,
    list_pickups_query,
)
from app.utils.pagination import get_pagination_params, paginate_query
from app.utils.responses import error_response, success_response

pickups_bp = Blueprint("pickups", __name__)


def _require_provider_profile():
    profile = g.current_user.provider_profile
    if profile is None:
        return None, error_response(
            "Complete your provider profile first.", "PROFILE_NOT_FOUND", 404
        )
    return profile, None


def _require_recipient_profile():
    profile = g.current_user.recipient_profile
    if profile is None:
        return None, error_response(
            "Complete your recipient profile first.", "PROFILE_NOT_FOUND", 404
        )
    return profile, None


@pickups_bp.route("", methods=["GET"])
@token_required
@role_required(UserRole.PROVIDER, UserRole.RECIPIENT)
def mine():
    """
    GET /api/pickups — every pickup involving me, as either the
    provider or the recipient (whichever profile(s) I have).

    Query parameters (optional): page, limit.
    """
    query = list_pickups_query(current_user=g.current_user)
    if query is None:
        return error_response(
            "Complete a provider or recipient profile first.",
            "PROFILE_NOT_FOUND",
            404,
        )

    page, limit = get_pagination_params(request.args)
    items, pagination = paginate_query(query, page, limit)
    return success_response(
        "Pickups retrieved successfully.",
        data={"pickups": [p.to_dict() for p in items], "pagination": pagination},
    )


@pickups_bp.route("/<int:pickup_id>", methods=["GET"])
@token_required
def get_one(pickup_id):
    """GET /api/pickups/<id> — owning provider or owning recipient only."""
    try:
        pickup = get_pickup_for_viewer(pickup_id=pickup_id, current_user=g.current_user)
    except PickupError as exc:
        return error_response(exc.message, exc.error_code, exc.status_code)

    return success_response("Pickup retrieved successfully.", data=pickup.to_dict())


@pickups_bp.route("/<int:pickup_id>/confirm", methods=["POST"])
@token_required
@role_required(UserRole.RECIPIENT)
def confirm(pickup_id):
    profile, err = _require_recipient_profile()
    if err:
        return err

    try:
        pickup = confirm_pickup(pickup_id=pickup_id, recipient_profile=profile)
    except PickupError as exc:
        return error_response(exc.message, exc.error_code, exc.status_code)

    return success_response("Pickup confirmed successfully.", data=pickup.to_dict())


@pickups_bp.route("/<int:pickup_id>/complete", methods=["POST"])
@token_required
@role_required(UserRole.PROVIDER)
def complete(pickup_id):
    """
    POST /api/pickups/<id>/complete

    Body (all optional):
        {"pickup_time": "2026-09-10T18:05:00+00:00", "confirmation_info": "Signed by J. Doe"}
    """
    profile, err = _require_provider_profile()
    if err:
        return err

    body = request.get_json(silent=True) or {}
    try:
        pickup = complete_pickup(pickup_id=pickup_id, provider_profile=profile, data=body)
    except PickupError as exc:
        return error_response(exc.message, exc.error_code, exc.status_code)

    return success_response("Pickup completed successfully.", data=pickup.to_dict())


@pickups_bp.route("/<int:pickup_id>/fail", methods=["POST"])
@token_required
@role_required(UserRole.PROVIDER)
def fail(pickup_id):
    """
    POST /api/pickups/<id>/fail

    Body (optional): {"reason": "Recipient did not show up."}
    """
    profile, err = _require_provider_profile()
    if err:
        return err

    body = request.get_json(silent=True) or {}
    try:
        pickup = fail_pickup(pickup_id=pickup_id, provider_profile=profile, data=body)
    except PickupError as exc:
        return error_response(exc.message, exc.error_code, exc.status_code)

    return success_response("Pickup marked as failed.", data=pickup.to_dict())


@pickups_bp.route("/<int:pickup_id>/cancel", methods=["POST"])
@token_required
@role_required(UserRole.RECIPIENT)
def cancel(pickup_id):
    profile, err = _require_recipient_profile()
    if err:
        return err

    try:
        pickup = cancel_pickup(pickup_id=pickup_id, recipient_profile=profile)
    except PickupError as exc:
        return error_response(exc.message, exc.error_code, exc.status_code)

    return success_response("Pickup cancelled successfully.", data=pickup.to_dict())

File: app/static/js/pickups.js
/**
 * FoodRescue Network — "My Pickups" dashboard (Phase 7)
 *
 * Drives templates/pickups/mine.html. Shared by both roles — a
 * provider and a recipient see the exact same list (GET /api/pickups
 * returns pickups scoped to whichever profile(s) the current user
 * has), just with different action buttons per card.
 *
 * ── Status filtering is client-side, on the current page only ──
 * GET /api/pickups only supports page/limit, not a status filter —
 * so this fetches a page (limit=50, the same ceiling the API already
 * clamps everything to) and filters what's already loaded, rather
 * than trying to fake server-side filtering. For the volume a single
 * provider/recipient deals with, one page covers everything in
 * practice; if that stops being true, the right fix is a `status`
 * query param on the API, not more cleverness here. The pagination
 * controls below page through the SERVER's pages (unfiltered) — the
 * status dropdown then narrows whatever's on the current page.
 *
 * ── Why we fetch listings/requests separately ──────────────────
 * PickupRecord.to_dict() only returns listing_id and request_id, not
 * the listing's food name or the request's quantity — so each is
 * fetched once per unique id (not once per pickup) and cached in a
 * map, the same N+1-but-bounded approach as Phase 6's dashboard.
 *
 * ── Phase 12 addition: success feedback ─────────────────────────
 * confirm/cancel/complete/fail previously only surfaced errors; a
 * successful action just silently re-rendered the list. showSuccess()
 * now gives the same treatment errors already had.
 */

const PICKUP_STATUS_LABELS = {
  scheduled: 'Scheduled',
  confirmed: 'Confirmed',
  completed: 'Completed',
  failed: 'Failed',
  cancelled: 'Cancelled',
};

// Reuse listing-status badge colors — nothing pickup-specific needed.
const PICKUP_BADGE_CLASS = {
  scheduled: 'pending',
  confirmed: 'reserved',
  completed: 'collected',
  failed: 'cancelled',
  cancelled: 'cancelled',
};

const ACTIVE_PICKUP_STATUSES = ['scheduled', 'confirmed'];

function escapeHtml(str) {
  const div = document.createElement('div');
  div.textContent = str == null ? '' : String(str);
  return div.innerHTML;
}

function formatDateTime(iso) {
  if (!iso) return 'Not yet set';
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return iso;
  return d.toLocaleString(undefined, {
    weekday: 'short', month: 'short', day: 'numeric',
    hour: 'numeric', minute: '2-digit',
  });
}

(function () {
  const root = document.querySelector('[data-page="my-pickups"]');
  if (!root) return;

  let currentPage = 1;
  let currentPagination = null;
  let currentPickups = [];
  let listingCache = {};  // listing_id -> listing dict
  let requestCache = {};  // request_id  -> request dict
  let myRole = null;      // 'provider' | 'recipient'

  function el(id) { return document.getElementById(id); }

  function showGate() {
    const user = Auth.isLoggedIn() ? Auth.getUser() : null;
    el('guest-notice').hidden = !!user;
    el('dashboard').hidden = true;
    el('no-profile-notice').hidden = true;
  }

  function showNoProfile() {
    el('dashboard').hidden = true;
    el('no-profile-notice').hidden = false;
  }

  function showError(message) {
    el('pk-alert').innerHTML =
      `<div class="alert alert--error">${escapeHtml(message)}</div>`;
  }

  function showSuccess(message) {
    el('pk-alert').innerHTML =
      `<div class="alert alert--success">${escapeHtml(message)}</div>`;
  }

  async function fetchUncached(ids, cache, fetchOne) {
    const missing = [...new Set(ids)].filter((id) => id != null && !(id in cache));
    if (!missing.length) return;
    const results = await Promise.all(missing.map(fetchOne));
    missing.forEach((id, i) => { cache[id] = results[i]; });
  }

  function cardHtml(pickup) {
    const statusLabel = PICKUP_STATUS_LABELS[pickup.status] || pickup.status;
    const badgeClass = PICKUP_BADGE_CLASS[pickup.status] || pickup.status;
    const listing = listingCache[pickup.listing_id];
    const req = requestCache[pickup.request_id];

    const title = listing ? escapeHtml(listing.food_name) : `Listing #${pickup.listing_id}`;
    const quantityLine = (req && listing)
      ? `<p class="text-muted">${escapeHtml(req.requested_quantity)} ${escapeHtml(listing.quantity_unit)}</p>`
      : '';
    const infoLine = pickup.confirmation_info
      ? `<p class="listing-card__desc">${escapeHtml(pickup.confirmation_info)}</p>`
      : '';

    let actions = '';
    if (myRole === 'recipient') {
      if (pickup.status === 'scheduled') {
        actions = `
          <button type="button" class="btn btn--primary" data-confirm-id="${pickup.id}">Confirm</button>
          <button type="button" class="btn btn--danger" data-cancel-id="${pickup.id}">Cancel</button>
        `;
      } else if (pickup.status === 'confirmed') {
        actions = `<button type="button" class="btn btn--danger" data-cancel-id="${pickup.id}">Cancel</button>`;
      }
    } else if (myRole === 'provider') {
      if (pickup.status === 'scheduled' || pickup.status === 'confirmed') {
        actions = `
          <button type="button" class="btn btn--primary" data-complete-id="${pickup.id}">Mark Completed</button>
          <button type="button" class="btn btn--danger" data-fail-id="${pickup.id}">Mark Failed</button>
        `;
      }
    }

    // Phase 8: a completed pickup has a permanent DistributionRecord
    // sitting on /distributions — point at it rather than leaving a
    // dead end once there's nothing left to action here.
    if (pickup.status === 'completed') {
      actions = `<a class="btn btn--secondary" href="/distributions">View in History</a>`;
    }

    return `
      <article class="card listing-card" data-pickup-card="${pickup.id}">
        <div class="listing-card__top">
          <span class="badge badge--${badgeClass}">${escapeHtml(statusLabel)}</span>
          <span class="text-muted listing-card__category">Pickup #${pickup.id}</span>
        </div>
        <h3><a href="/listings/${pickup.listing_id}">${title}</a></h3>
        ${quantityLine}
        <p class="text-muted">Pickup time: ${escapeHtml(formatDateTime(pickup.pickup_time))}</p>
        ${infoLine}
        <div class="listing-card__actions">${actions}</div>
      </article>
    `;
  }

  function render() {
    const filter = el('status-filter').value;
    const gridEl = el('pk-grid');
    const emptyEl = el('pk-empty');

    const shown = filter === 'active'
      ? currentPickups.filter((p) => ACTIVE_PICKUP_STATUSES.includes(p.status))
      : filter
        ? currentPickups.filter((p) => p.status === filter)
        : currentPickups;

    if (!shown.length) {
      gridEl.hidden = true;
      gridEl.innerHTML = '';
      emptyEl.hidden = false;
      emptyEl.textContent = filter
        ? `No pickups with that status on this page.`
        : 'No pickups yet.';
      return;
    }

    emptyEl.hidden = true;
    gridEl.innerHTML = shown.map(cardHtml).join('');
    gridEl.hidden = false;

    gridEl.querySelectorAll('[data-confirm-id]').forEach((btn) => {
      btn.addEventListener('click', () => act(btn.dataset.confirmId, 'confirm', btn, 'Confirm'));
    });
    gridEl.querySelectorAll('[data-cancel-id]').forEach((btn) => {
      btn.addEventListener('click', () => act(btn.dataset.cancelId, 'cancel', btn, 'Cancel'));
    });
    gridEl.querySelectorAll('[data-complete-id]').forEach((btn) => {
      btn.addEventListener('click', () => act(btn.dataset.completeId, 'complete', btn, 'Mark Completed'));
    });
    gridEl.querySelectorAll('[data-fail-id]').forEach((btn) => {
      btn.addEventListener('click', () => act(btn.dataset.failId, 'fail', btn, 'Mark Failed'));
    });
  }

  const ACT_SUCCESS_MESSAGES = {
    confirm: 'Pickup confirmed.',
    cancel: 'Pickup cancelled.',
    complete: 'Pickup marked as completed.',
    fail: 'Pickup marked as failed.',
  };

  async function act(pickupId, action, btn, label) {
    const verbs = {
      confirm: 'Confirm this pickup time?',
      cancel: 'Cancel this pickup? The request will revert to accepted so it can be rescheduled.',
      complete: 'Mark this pickup as completed? This records the handover permanently.',
      fail: 'Mark this pickup as failed (e.g. a no-show)? The request will revert to accepted.',
    };
    if (!confirm(verbs[action] || `${label}?`)) return;

    el('pk-alert').innerHTML = '';
    btn.disabled = true;
    btn.textContent = 'Working…';

    try {
      await Api.post(`/api/pickups/${pickupId}/${action}`);
      await load({ silent: true });
      showSuccess(ACT_SUCCESS_MESSAGES[action] || `${label} done.`);
    } catch (err) {
      btn.disabled = false;
      btn.textContent = label;

      if (err.status === 401) {
        if (typeof renderAuthArea === 'function') renderAuthArea();
        showGate();
        return;
      }
      showError(err.message || `Could not ${label.toLowerCase()}. Please try again.`);
      const card = document.querySelector(`[data-pickup-card="${pickupId}"]`);
      if (card) card.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
    }
  }

  async function load(options = {}) {
    const loadingEl = el('pk-loading');
    const gridEl = el('pk-grid');
    const paginationEl = el('pagination');

    if (!options.silent) {
      loadingEl.hidden = false;
      gridEl.hidden = true;
      el('pk-empty').hidden = true;
      paginationEl.hidden = true;
    }

    try {
      const result = await Api.get(`/api/pickups?page=${currentPage}&limit=50`);
      currentPickups = result.data.pickups;
      currentPagination = result.data.pagination;

      await Promise.all([
        fetchUncached(
          currentPickups.map((p) => p.listing_id),
          listingCache,
          (id) => Api.get(`/api/listings/${id}`).then((r) => r.data).catch(() => null)
        ),
        fetchUncached(
          currentPickups.map((p) => p.request_id),
          requestCache,
          (id) => Api.get(`/api/requests/${id}`).then((r) => r.data).catch(() => null)
        ),
      ]);

      loadingEl.hidden = true;
      render();

      if (currentPagination.total_pages > 1) {
        paginationEl.hidden = false;
        el('page-info').textContent =
          `Page ${currentPagination.page} of ${currentPagination.total_pages} (${currentPagination.total} total)`;
        el('prev-page').disabled = currentPagination.page <= 1;
        el('next-page').disabled = currentPagination.page >= currentPagination.total_pages;
      }
    } catch (err) {
      loadingEl.hidden = true;
      if (err.status === 401) {
        if (typeof renderAuthArea === 'function') renderAuthArea();
        showGate();
        return;
      }
      if (err.code === 'PROFILE_NOT_FOUND') {
        showNoProfile();
        return;
      }
      showError(err.message || 'Could not load your pickups. Please try again.');
    }
  }

  document.addEventListener('DOMContentLoaded', () => {
    const user = Auth.isLoggedIn() ? Auth.getUser() : null;
    if (!user) {
      showGate();
      return;
    }
    if (user.role !== 'provider' && user.role !== 'recipient') {
      showNoProfile();
      return;
    }

    myRole = user.role;
    el('dashboard').hidden = false;
    el('status-filter').addEventListener('change', render);
    el('prev-page').addEventListener('click', () => {
      if (currentPage > 1) { currentPage -= 1; load(); }
    });
    el('next-page').addEventListener('click', () => {
      if (currentPagination && currentPage < currentPagination.total_pages) {
        currentPage += 1; load();
      }
    });

    load();
  });
})();